protocols/consensus-gossip/tuple-space.md

```diff
@@ -0,0 +1,192 @@
+
+#### State Validity
+
+Let $\text{ValSet}_h$ be the set of validators of height $h$ and $\text{Prop}_{h,r}$ be the proposal of round $r$ of height $h$.
+
+Given that each validator can execute each step only once per round, a query that specifies height, round, step and validator must either return empty or a single tuple.
+
+- $\forall h \in \N, r \in \N, s \in \{\text{Proposal, PreVote, PreCommit}\}, v \in \text{ValSet}_h$,  $\lang h, r, s, v, * \rang$ returns at most one value.
+
+In the specific case of the Proposal step, only the proposer of the round can have a matching entry.
+
+- $\forall h \in \N, r \in \N, \lang h, r, \text{Proposal}, *, * \rang$ returns at most one value and it also matches $\lang h, r, \text{Proposal}, \text{Prop}_{h,r}, * \rang$.
+
+A violation of these rules is a proof of misbehavior.
+
+
+### Eventual Convergence
+
+Consider the following definition for **Eventual Convergence**.
+
+|Eventual Convergence|
+|-----|
+| If there exists a correct process $p$ such that $e \in t_p$, then, eventually, for every correct process $q$, $e \in t_q$ or there exists a correct process $r$ such that $\bar{e} \in t_r$.|
+
+> **Note**
+> Nodes may learn of an entry deletion before learning of its addition.
+
+In order to ensure convergence even in the presence of failures, the network must be connected in such a way to allow communication around any malicious nodes and to provide redundant paths between correct ones.
+This can be achieved if there is a GST, after which timeouts eventually do not expire precociously, given that they all can be adjusted to reasonable values, which implies that all communication will eventually happen timely, which implies that the tuple space will converge and keep converging.
+Formally, if there is a GST then following holds true:
+
+| Eventual $\Delta$-Timely Convergence |
+|---|
+| If $e\in t_p$, for some correct process $p$, at instant $t$, then by $\text{max}(t,\text{GST}) + \Delta$, either $e \in t_q$, for every correct process $q$ or $\bar{e} \in t_p$.
+
+Although GST may be too strong an expectation, in practice timely communication frequently happens within small stable periods, also leading to convergence.
+
+
+### Why use a Tuple Space
+
+Let's recall why we are considering using a tuple space to propagate Tendermint's messages.
+It should be straightforward to see that Reliable Broadcast may be used to achieve Eventual Convergence and Gossip Communication may be used to implement Eventual $\Delta$-Timely Convergence:
+to add an entry to the tuple space, broadcast the entry;
+once delivered, add the entry to the local view.[^proof]
+If indeed we use Gossip Communication, then there are no obvious gains.
+
+It should also be clear that if no entries are ever removed from the tuple space, then the inverse is also true:
+to broadcast a message, add it to the local view;
+once an entry is added to the local view, deliver it.
```

Comment:
The latter is probably not true.

Assume a Byzantine process that equivocates. Lets $p$ be a correct process that receives a version of the equivocating message, and $q$ a correct process that receives the other version. From Gossip communication (ii), any correct process should deliver both versions. Eventual $\Delta$-Timely Convergence only allows us to deliver one of them.

Reply:
Really nice catch.
I need to think about how to fix this because the goal here is to be weaker than Gossip Communication, not stronger.